### PR TITLE
feat: adds parallelization to `SenderRecovery` and `TxLookup` using `SnapshotProvider`

### PR DESCRIPTION
**Pull request by joshieDo**
### Changes
* closes https://github.com/paradigmxyz/reth/issues/6249 
* make `transaction_hashes_by_range` on static files parallel: each rayon worker reads the file(and deserializes) on their own vs mdbx where reading and deserialization is done on the main thread before sending it to rayon workers for hashing.
* SenderRecovery reads from snapshot provider in parallel as well.
* `chunk_size` on parallelization is set to `100`: Kinda arbitrary. The way it was before it was too high and there were a bunch of rayon threads waiting for work. (image comparison incoming)

### Benches
holesky `maxperf` @ block 925555

`TransactionLookup` went from [57s-59s] to [47s-50s].
`SenderRecovery` didn't benefit from static files (heavily cpu bound), but the `chunk_size` reduction helped it go from 49s to 44s.

### CODE REVIEW

File: bin/reth/src/commands/stage/run.rs [hunk 1]
```diff
@@ -269,6 +269,8 @@ impl Command {
             checkpoint: Some(checkpoint.with_block_number(self.from)),
         };
 
+        let start = Instant::now();
+        info!(target: "reth::cli", stage= ?self.stage, "Executing stage.", );
```
Comment (mattsse) [inline bin/reth/src/commands/stage/run.rs:273]:
```suggestion
        info!(target: "reth::cli", stage= ?self.stage, "Executing stage.");
```

File: bin/reth/src/commands/stage/run.rs [hunk 2]
```diff
@@ -284,6 +286,7 @@ impl Command {
                 break
             }
         }
+        info!(target: "reth::cli", stage= ?self.stage, time = ?start.elapsed().as_secs(), "Finished stage.");
```
Comment (mattsse) [inline bin/reth/src/commands/stage/run.rs:289]: the debug impls already formats properly
```suggestion
        info!(target: "reth::cli", stage= ?self.stage, time = ?start.elapsed(), "Finished stage.");
```

File: crates/storage/provider/src/providers/snapshot/manager.rs [hunk 3]
```diff
@@ -29,18 +30,39 @@ use reth_primitives::{
...
+    sync::{mpsc, Arc},
 };
 
 /// Alias type for a map that can be queried for block ranges from a transaction
 /// segment respectively. It uses `TxNumber` to represent the transaction end of a snapshot range.
 type SegmentRanges = HashMap<SnapshotSegment, BTreeMap<TxNumber, SegmentRangeInclusive>>;
 
 /// [`SnapshotProvider`] manages all existing [`SnapshotJarProvider`].
+#[derive(Debug, Default, Clone)]
+pub struct SnapshotProvider(Arc<SnapshotProviderInner>);
```
Comment (mattsse) [inline crates/storage/provider/src/providers/snapshot/manager.rs:44]: cool, this now closes the other pr that adds this main

File: crates/storage/provider/src/providers/snapshot/manager.rs [hunk 4]
```diff
@@ -69,14 +91,16 @@ impl SnapshotProvider {
...
-
+}
+impl SnapshotProvider {
     /// Loads filters into memory when creating a [`SnapshotJarProvider`].
-    pub fn with_filters(mut self) -> Self {
-        self.load_filters = true;
-        self
+    pub fn with_filters(self) -> Self {
+        let mut provider =
+            Arc::try_unwrap(self.0).expect("should be called when initializing only.");
```
Comment (mattsse) [inline crates/storage/provider/src/providers/snapshot/manager.rs:101]: for followup, this should be called during SnapshotProvider::new

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, nits

pending @shekhirin

File: bin/reth/src/commands/stage/run.rs [hunk 5]
```diff
@@ -269,6 +269,8 @@ impl Command {
             checkpoint: Some(checkpoint.with_block_number(self.from)),
         };
 
+        let start = Instant::now();
+        info!(target: "reth::cli", stage= ?self.stage, "Executing stage.");
```
Comment (shekhirin) [inline bin/reth/src/commands/stage/run.rs:273]: nit
```suggestion
        info!(target: "reth::cli", stage = %self.stage, "Executing stage");
```

File: bin/reth/src/commands/stage/run.rs [hunk 6]
```diff
@@ -284,6 +286,7 @@ impl Command {
                 break
             }
         }
+        info!(target: "reth::cli", stage= ?self.stage, time = ?start.elapsed(), "Finished stage.");
```
Comment (shekhirin) [inline bin/reth/src/commands/stage/run.rs:289]: nit
```suggestion
        info!(target: "reth::cli", stage = %self.stage, time = ?start.elapsed(), "Finished stage");
```

File: crates/stages/src/stages/sender_recovery.rs [hunk 7]
```diff
@@ -83,43 +82,49 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
...
-        // prevents an edge case
-        // where the chunk size is either 0 or too small
-        // to gain anything from using more than 1 thread
-        let chunk_size = chunk_size.max(16);
-
-        for chunk in &tx_range.zip(transactions).chunks(chunk_size) {
+        // Transactions are different size, so chunks will not all take the same processing time. If
+        // chunks are too big, there will be idle threads waiting for work. Choosing an
+        // arbitrary smaller value to make sure it doesn't happen.
+        let chunk_size = 100;
```
Comment (shekhirin) [inline crates/stages/src/stages/sender_recovery.rs:94]: hmm, shouldn't we somehow depend on the number of available threads anyway?

File: crates/stages/src/stages/sender_recovery.rs [hunk 8]
```diff
@@ -83,43 +82,49 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
...
+        for chunk_range in chunks {
             // An _unordered_ channel to receive results from a rayon job
             let (recovered_senders_tx, recovered_senders_rx) = mpsc::channel();
             channels.push(recovered_senders_rx);
-            // Note: Unfortunate side-effect of how chunk is designed in itertools (it is not Send)
-            let chunk: Vec<_> = chunk.collect();
 
-            // Spawn the sender recovery task onto the global rayon pool
-            // This task will send the results through the channel after it recovered the senders.
+            let manager = provider.snapshot_provider().clone();
```
Comment (shekhirin) [inline crates/stages/src/stages/sender_recovery.rs:107]: nit, a bit confusing

```suggestion
            let snapshot_provider = provider.snapshot_provider().clone();
```

File: crates/stages/src/test_utils/test_db.rs [hunk 9]
```diff
@@ -194,18 +194,26 @@ impl TestStageDB {
     /// Insert ordered collection of [SealedBlock] into corresponding tables.
     /// Superset functionality of [TestStageDB::insert_headers].
     ///
+    /// If tx_offset is set to `None`, then transactions will be stored on static files, otherwise
```
Comment (shekhirin) [inline crates/stages/src/test_utils/test_db.rs:197]: oof imo it's easy to misuse. Maybe an explicit enum?

**Review by shekhirin — APPROVED** [review]
LGTM, only nits

File: crates/stages/src/stages/sender_recovery.rs (hunk 7 above)
Comment (joshieDo) [reply crates/stages/src/stages/sender_recovery.rs:94]: tbh it's hard to get the heuristic right i think. 

less is more in this case, thats my original point, but open to suggestions

**Review by shekhirin — APPROVED** [review]

**Pull request merged**